Polyfills/Window/Source/TimeoutDispatcher.cpp

```diff
@@ -0,0 +1,160 @@
+#include "TimeoutDispatcher.h"
+
+#include <cassert>
+
+namespace Babylon::Polyfills::Internal
+{
+    namespace
+    {
+        using TimePoint = std::chrono::time_point<std::chrono::steady_clock, std::chrono::microseconds>;
+
+        TimePoint Now()
+        {
+            return std::chrono::time_point_cast<std::chrono::microseconds, std::chrono::steady_clock>(std::chrono::steady_clock::now());
+        }
+    }
+
+    struct TimeoutDispatcher::Timeout
+    {
+        TimeoutId id;
+
+        // Make this non-shared when JsRuntime::Dispatch supports it.
+        std::shared_ptr<Napi::FunctionReference> function;
+
+        TimePoint time;
+
+        Timeout(TimeoutId id, Napi::Function function, TimePoint time)
+            : id{id}
+            , function{std::make_shared<Napi::FunctionReference>(Napi::Persistent(function))}
+            , time{time}
+        {
+        }
+
+        Timeout(const Timeout&) = delete;
+        Timeout(Timeout&&) = delete;
+    };
+
+    TimeoutDispatcher::TimeoutDispatcher(Babylon::JsRuntime& runtime)
+        : m_runtime(runtime)
+    {
+        m_thread = std::thread{&TimeoutDispatcher::ThreadFunction, this};
+    }
+
+    TimeoutDispatcher::~TimeoutDispatcher()
+    {
+        {
+            std::unique_lock<std::mutex> lk{m_mutex};
+            m_idMap.clear();
+            m_timeMap.clear();
+            m_shutdown = true;
+            m_condVariable.notify_one();
+        }
+
+        m_thread.join();
+    }
+
+    TimeoutDispatcher::TimeoutId TimeoutDispatcher::Dispatch(Napi::Function function, std::chrono::milliseconds delay)
+    {
+        if (delay.count() < 0)
+        {
+            delay = std::chrono::milliseconds{0};
+        }
+
+        std::unique_lock<std::mutex> lk{m_mutex};
+
+        const auto earliestTime = m_timeMap.empty() ? TimePoint::max()
+                                                    : m_timeMap.cbegin()->second->time;
+        const auto id = NextTimeoutId();
+        const auto time = Now() + delay;
+        const auto result = m_idMap.insert({id, std::make_unique<Timeout>(id, std::move(function), time)});
+        m_timeMap.insert({time, result.first->second.get()});
+
+        if (time <= earliestTime)
+        {
+            m_runtime.Dispatch([this](Napi::Env){ m_condVariable.notify_one(); });
```

Comment:
I think there is a problem here. `m_condVariable.notify_one();` is not under the `std::unique_lock` scope anymore, so this is not correct usage of condition variables.

Reply:
I'm also not exactly sure what this mutex is supposed to protect. All of the members or just the condition variable?

Reply:
Maybe all of this code under the lock should be on the JS thread?

Reply:
Oof! Good catch. Ya, I'll try moving it all to the JS thread. Can remove the mutex and make `m_shutdown` a `bool` again instead of a `std::atomic<bool>` if we do it that way?

Reply:
I don't think so. `m_shutdown` is still accessed from the Timeout thread.

Reply:
Like this?
```C++
    TimeoutDispatcher::TimeoutId TimeoutDispatcher::Dispatch(Napi::Function function, std::chrono::milliseconds delay)
    {
        if (delay.count() < 0)
        {
            delay = std::chrono::milliseconds{0};
        }

        const auto id = NextTimeoutId();

        m_runtime.Dispatch([this, function, delay, id](Napi::Env){
            std::unique_lock<std::mutex> lk{m_mutex};

            const auto earliestTime = m_timeMap.empty() ? TimePoint::max()
                                                        : m_timeMap.cbegin()->second->time;
            const auto time = Now() + delay;
            const auto result = m_idMap.insert({id, std::make_unique<Timeout>(id, std::move(function), time)});
            m_timeMap.insert({time, result.first->second.get()});

            if (time <= earliestTime)
            {
                m_condVariable.notify_one();
            }
        });

        return id;
    }

    void TimeoutDispatcher::Clear(TimeoutId id)
    {
        m_runtime.Dispatch([this, id](Napi::Env){
            std::unique_lock<std::mutex> lk{m_mutex};
            const auto itId = m_idMap.find(id);
            if (itId != m_idMap.end())
            {
                const auto& timeout = itId->second;
                const auto timeRange = m_timeMap.equal_range(timeout->time);

                assert(timeRange.first != m_timeMap.end() && "m_idMap and m_timeMap are out of sync");

                for (auto itTime = timeRange.first; itTime != timeRange.second; itTime++)
                {
                    if (itTime->second->id == id)
                    {
                        m_timeMap.erase(itTime);
                        break;
                    }
                }

                m_idMap.erase(itId);
            }
        });
    }
```

Reply:
Seems right.